File: packages/python/src/sah_sdk/core/session.py
"""Steady-state session module + transparent auto-refresh interceptor.

Pairs with :mod:`sah_sdk.core.bootstrap` (the byte-pinned HMAC algorithm)
and :func:`sah_sdk.core.sign_request.sign_request` (per-request signing).

Consumer view: never call :func:`SessionStore.bootstrap` directly. The
:class:`SessionInterceptor` wraps the underlying :class:`Transport` and
fires bootstrap on miss/expiry, retries once on refreshable 401 auth errors,
and collapses concurrent cold-start callers onto a single round-trip via
a cached :class:`threading.Lock` + monotonically-stored result.

Concurrency:
    - :class:`threading.RLock` guards the cached :class:`Session`.
    - Single-flight is enforced by holding the lock for the duration of
      the in-flight bootstrap, so the second arriving thread observes
      the freshly-stored session on lock acquisition and skips the
      network round-trip.
    - The 30-second grace overlap lives server-side
      (``services/account/internal/handler/sessions_bootstrap.go``);
      the client just retries on 401 and never tracks the previous
      secret.
"""

from __future__ import annotations

import json
import threading
from collections.abc import Callable
from dataclasses import dataclass
from datetime import datetime, timedelta, timezone
from urllib.parse import urlparse

from .bootstrap import BootstrapInput, build_bootstrap_signature
from .sign_request import sign_request
from .transport import Transport, TransportResponse


@dataclass(frozen=True, slots=True)
class Session:
    """Cached credential bundle returned by ``POST /v1/sessions``.

    The ``secret`` is the HMAC key used to sign every steady-state
    request. Treat as a credential: never log, never persist to disk.
    """

    session_id: str
    session_secret: str
    expires_at: datetime


@dataclass(frozen=True, slots=True)
class ExchangeInput:
    """Inputs needed to bootstrap a session."""

    keycode: str
    email: str
    license_key: str
    device_id: str


class SessionExchangeError(RuntimeError):
    """Raised when ``POST /v1/sessions`` returns non-2xx."""


SessionClock = Callable[[], datetime]
_SESSION_EXPIRED_CODE = "session_expired"
_UNAUTHORIZED_CODE = "unauthorized"


def _system_clock() -> datetime:
    return datetime.now(tz=timezone.utc)


class SessionStore:
    """Thread-safe atomic session cache + single-flight bootstrap driver.

    One :class:`SessionStore` per SDK client. Safe for concurrent use
    from any thread.
    """

    #: How close to expiry :meth:`on_activity` proactively re-mints.
    PROACTIVE_WINDOW = timedelta(minutes=5)
    #: Minimum delay before retrying after a failed session exchange.
    FAILURE_BACKOFF = timedelta(seconds=1)

    def __init__(
        self,
        *,
        transport: Transport,
        base_url: str,
        exchange_input: ExchangeInput,
        clock: SessionClock | None = None,
    ) -> None:
        if not base_url:
            raise ValueError("SessionStore: base_url must be a non-empty string")
        if not (
            exchange_input.keycode
            and exchange_input.email
            and exchange_input.license_key
            and exchange_input.device_id
        ):
            raise ValueError(
                "SessionStore: ExchangeInput requires non-empty keycode, "
                "email, license_key, device_id"
            )
        self._transport = transport
        self._base_url = base_url.rstrip("/")
        self._input = exchange_input
        self._clock = clock or _system_clock
        self._lock = threading.RLock()
        self._current: Session | None = None
        self._exchange_blocked_until: datetime | None = None
        self._last_exchange_error: Exception | None = None
        # Single-flight counter for tests. Incremented inside the
        # critical section, so reads under the same lock are consistent.
        self._bootstrap_count = 0

    @property
    def bootstrap_count(self) -> int:
        """Number of network bootstraps the store has executed.

        Tests assert exactly 1 after concurrent cold-start calls — the
        single-flight invariant.
        """
        with self._lock:
            return self._bootstrap_count

    def current_secret(self) -> Session | None:
        """Return the cached session if present and not past expiry.

        Returns ``None`` if the caller must bootstrap.
        """
        with self._lock:
            if self._current is None:
                return None
            if self._clock() >= self._current.expires_at:
                return None
            return self._current

    def bootstrap(self) -> Session:
        """Perform ``POST /v1/sessions`` with the embedded HMAC signature.

        Single-flight: concurrent callers serialize on ``self._lock``;
        the second arriver observes the freshly-cached session and
        skips the network call.
        """
        with self._lock:
            now = self._clock()
            # Double-checked: a concurrent caller may have already
            # bootstrapped while we waited for the lock.
            if self._current is not None and now < self._current.expires_at:
                return self._current
            self._raise_recent_exchange_error(now)
            return self._exchange_locked(now)

    def invalidate(self) -> None:
        """Clear the cached session. Called by the interceptor on 401."""
        with self._lock:
            self._current = None

    def _invalidate_if_current(self, session_id: str) -> None:
        with self._lock:
            if self._current is not None and self._current.session_id == session_id:
                self._current = None

    def _refresh_if_current(self, session_id: str) -> Session:
        with self._lock:
            now = self._clock()
            if self._current is not None and self._current.session_id != session_id:
                return self._current
            self._raise_recent_exchange_error(now)
            return self._exchange_locked(now)

    def on_activity(self) -> None:
        """Consumer-facing proactive-refresh hook.

        If the cached session is within :attr:`PROACTIVE_WINDOW` of
        expiry, re-mint now so the next product call doesn't pay the
        bootstrap round-trip. Safe to call on every consumer-side
        activity boundary (button click, IPC tick).
        """
        with self._lock:
            now = self._clock()
            cur = self._current
            if cur is None or now + self.PROACTIVE_WINDOW >= cur.expires_at:
                self._raise_recent_exchange_error(now)
                self._exchange_locked(now)

    def _exchange_locked(self, now: datetime) -> Session:
        try:
            sess = self._do_exchange()
        except Exception as exc:
            self._exchange_blocked_until = now + self.FAILURE_BACKOFF
            self._last_exchange_error = exc
            raise
        self._current = sess
        self._exchange_blocked_until = None
        self._last_exchange_error = None
        self._bootstrap_count += 1
        return sess

    def _raise_recent_exchange_error(self, now: datetime) -> None:
        if (
            self._last_exchange_error is not None
            and self._exchange_blocked_until is not None
            and now < self._exchange_blocked_until
        ):
            raise self._last_exchange_error

    def _do_exchange(self) -> Session:
        ts = int(self._clock().timestamp())
        sig = build_bootstrap_signature(
            BootstrapInput(
                keycode=self._input.keycode,
                email=self._input.email,
                license_key=self._input.license_key,
                device_id=self._input.device_id,
                method="POST",
                path="/v1/sessions",
                timestamp=ts,
            )
        )
        resp = self._transport.request(
            "POST",
            self._base_url + "/v1/sessions",
            headers={
                "Content-Type": "application/json",
                "X-Device-ID": self._input.device_id,
                "ISA-Signature": f"t={ts},v1={sig.hex}",
            },
            body=sig.serialized_body,
        )
        if not 200 <= resp.status < 300:
            raise SessionExchangeError(
                f"POST /v1/sessions returned {resp.status}: {resp.body[:200]}"
            )
        payload = json.loads(resp.body)["data"]
        return Session(
            session_id=payload["sessionId"],
            session_secret=payload["sessionSecret"],
            expires_at=datetime.fromisoformat(
                payload["expiresAt"].replace("Z", "+00:00")
            ),
        )


class SessionInterceptor:
    """Transparent :class:`Transport` wrapper.

    Sits between product namespaces (``zyins``, ``account``, ``rapidsign``,
    ``proxy``) and the underlying :class:`HttpTransport`. Wiring at the
    transport seam means every existing product method inherits
    auto-refresh without per-method changes.

    Behavior:
        1. On every :meth:`request`, read :meth:`SessionStore.current_secret`.
           If ``None``, call :meth:`SessionStore.bootstrap` (single-flight).
        2. Sign the request via :func:`sign_request`.
        3. Forward to the inner transport.
        4. On refreshable 401 auth errors, bootstrap and replay once. A
           second 401 is returned unchanged.
    """

    def __init__(self, *, store: SessionStore, inner: Transport) -> None:
        if store is None:
            raise ValueError("SessionInterceptor: store is required")
        if inner is None:
            raise ValueError("SessionInterceptor: inner transport is required")
        self._store = store
        self._inner = inner

    def request(
        self,
        method: str,
        url: str,
        *,
        headers: dict[str, str],
        body: str | None = None,
    ) -> TransportResponse:
        resp, session = self._sign_and_send(method, url, headers, body)
        code = _auth_error_code(resp)
        if not self._should_refresh(code):
            return resp
        try:
            if code == _SESSION_EXPIRED_CODE:
                self._store._invalidate_if_current(session.session_id)
            self._store._refresh_if_current(session.session_id)
        except Exception:
            return resp
        retry, _ = self._sign_and_send(method, url, headers, body)
        return retry

    def _should_refresh(self, code: str | None) -> bool:
        return code in (_SESSION_EXPIRED_CODE, _UNAUTHORIZED_CODE)

    def _sign_and_send(
        self,
        method: str,
        url: str,
        headers: dict[str, str],
        body: str | None,
    ) -> tuple[TransportResponse, Session]:
        sess = self._store.current_secret() or self._store.bootstrap()
        path = _path_of(url)
        signed = sign_request(
            method=method,
            path=path,
            body=body or "",
            session_id=sess.session_id,
            session_secret=sess.session_secret,
        )
        merged = dict(headers)
        merged.update(signed.as_dict())
        return self._inner.request(method, url, headers=merged, body=body), sess


def _auth_error_code(resp: TransportResponse) -> str | None:
    if resp.status != 401:
        return None
    ct = resp.headers.get("Content-Type", "") or resp.headers.get("content-type", "")
    if "json" not in ct:
        return None
    try:
        payload = json.loads(resp.body)
    except (ValueError, TypeError):
        return None
    if not isinstance(payload, dict):
        return None
    code = payload.get("code")
    return code if isinstance(code, str) else None


def _path_of(url: str) -> str:
    if "://" not in url:
        return url.split("?", 1)[0]
    return urlparse(url).path or "/"

File: packages/csharp/src/Core/Session.cs
// Steady-state session module + transparent auto-refresh interceptor.
//
// Pairs with Bootstrap.Build (the byte-pinned HMAC algorithm) and
// SignRequest.Sign (per-request signing).
//
// Consumer view: never call SessionStore.BootstrapAsync directly. The
// SessionInterceptor wraps an inner ITransport and fires bootstrap on
// miss/expiry, retries once on 401 session_expired, and collapses
// concurrent cold-start callers onto a single round-trip via a
// SemaphoreSlim + double-checked cache.
//
// The 30-second grace overlap lives server-side
// (services/account/internal/handler/sessions_bootstrap.go); the
// client just retries on 401 and never tracks the previous secret.

using System.Globalization;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace Isa.Sdk.Core;

/// <summary>
/// Cached credential bundle returned by POST /v1/sessions. The Secret
/// is the HMAC key used to sign every steady-state request — treat as
/// a credential: never log, never persist beyond memory.
/// </summary>
public sealed record Session(
    string SessionId,
    string SessionSecret,
    DateTimeOffset ExpiresAt);

/// <summary>Inputs the bootstrap signature needs.</summary>
public sealed record SessionExchangeInput(
    string Keycode,
    string Email,
    string LicenseKey,
    string DeviceId);

/// <summary>Thrown when POST /v1/sessions returns non-2xx.</summary>
public sealed class SessionExchangeException : Exception
{
    public int Status { get; }

    public SessionExchangeException(int status, string message) : base(message)
    {
        Status = status;
    }
}

/// <summary>
/// Atomic session cache + single-flight bootstrap driver.
///
/// Thread-safe; one <see cref="SessionStore"/> per SDK client.
/// <see cref="SemaphoreSlim"/>(1,1) guards the bootstrap critical
/// section; a double-checked read of <see cref="CurrentSecret"/>
/// inside the semaphore collapses concurrent cold-start callers onto
/// the same round-trip.
/// </summary>
public sealed class SessionStore : IDisposable
{
    /// <summary>How close to expiry <see cref="OnActivityAsync"/> proactively re-mints.</summary>
    public static readonly TimeSpan ProactiveWindow = TimeSpan.FromMinutes(5);

    private readonly ITransport _transport;
    private readonly IClock _clock;
    private readonly string _baseUrl;
    private readonly SessionExchangeInput _input;
    private readonly SemaphoreSlim _gate = new(1, 1);
    private Session? _current;
    private int _bootstrapCount;

    public SessionStore(
        ITransport transport,
        string baseUrl,
        SessionExchangeInput input,
        IClock? clock = null)
    {
        if (transport is null)
        {
            throw new ArgumentNullException(nameof(transport));
        }
        if (input is null)
        {
            throw new ArgumentNullException(nameof(input));
        }
        if (string.IsNullOrWhiteSpace(baseUrl))
        {
            throw new ArgumentException(
                "SessionStore: baseUrl must be non-empty",
                nameof(baseUrl));
        }
        if (string.IsNullOrWhiteSpace(input.Keycode) ||
            string.IsNullOrWhiteSpace(input.Email) ||
            string.IsNullOrWhiteSpace(input.LicenseKey) ||
            string.IsNullOrWhiteSpace(input.DeviceId))
        {
            throw new ArgumentException(
                "SessionStore: ExchangeInput requires non-empty keycode, email, licenseKey, deviceId",
                nameof(input));
        }
        _transport = transport;
        _clock = clock ?? SystemClock.Instance;
        _baseUrl = baseUrl.TrimEnd('/');
        _input = input;
    }

    /// <summary>
    /// Number of network bootstraps performed. Tests assert exactly 1
    /// after 10 concurrent product calls — the single-flight invariant.
    /// </summary>
    public int BootstrapCount => Volatile.Read(ref _bootstrapCount);

    /// <summary>Return the cached session if present and not past expiry.</summary>
    public Session? CurrentSecret()
    {
        var cur = Volatile.Read(ref _current);
        if (cur is null)
        {
            return null;
        }
        if (_clock.UtcNow() >= cur.ExpiresAt)
        {
            return null;
        }
        return cur;
    }

    /// <summary>
    /// Perform POST /v1/sessions with the embedded HMAC signature.
    /// Concurrent callers serialize on a SemaphoreSlim; the second
    /// arriver observes the freshly-cached session and skips the
    /// network call.
    /// </summary>
    public async Task<Session> BootstrapAsync(CancellationToken ct = default)
    {
        return await BootstrapAsync(forceRefresh: false, ct).ConfigureAwait(false);
    }

    private async Task<Session> BootstrapAsync(bool forceRefresh, CancellationToken ct)
    {
        await _gate.WaitAsync(ct).ConfigureAwait(false);
        try
        {
            var cur = _current;
            if (!forceRefresh && cur is not null && _clock.UtcNow() < cur.ExpiresAt)
            {
                return cur;
            }
            var sess = await DoExchangeAsync(ct).ConfigureAwait(false);
            Volatile.Write(ref _current, sess);
            Interlocked.Increment(ref _bootstrapCount);
            return sess;
        }
        finally
        {
            _gate.Release();
        }
    }

    /// <summary>Clear the cached session. Called by the interceptor on 401.</summary>
    public void Invalidate()
    {
        Volatile.Write(ref _current, null);
    }

    /// <summary>
    /// Consumer-facing proactive-refresh hook. If the cached session
    /// is within <see cref="ProactiveWindow"/> of expiry, re-mint now.
    /// </summary>
    public async Task OnActivityAsync(CancellationToken ct = default)
    {
        var cur = Volatile.Read(ref _current);
        if (cur is null)
        {
            await BootstrapAsync(ct).ConfigureAwait(false);
            return;
        }
        if (_clock.UtcNow() + ProactiveWindow >= cur.ExpiresAt)
        {
            await BootstrapAsync(forceRefresh: true, ct).ConfigureAwait(false);
        }
    }

    private async Task<Session> DoExchangeAsync(CancellationToken ct)
    {
        var ts = _clock.UtcNow().ToUnixTimeSeconds();
        var sig = Bootstrap.Build(new BootstrapInput(
            Keycode: _input.Keycode,
            Email: _input.Email,
            LicenseKey: _input.LicenseKey,
            DeviceId: _input.DeviceId,
            Method: "POST",
            Path: "/v1/sessions",
            Timestamp: ts));
        var headers = new Dictionary<string, string>
        {
            ["Content-Type"] = "application/json",
            ["X-Device-ID"] = _input.DeviceId,
            ["ISA-Signature"] = string.Format(
                CultureInfo.InvariantCulture, "t={0},v1={1}", ts, sig.Hex),
        };
        var req = new TransportRequest(
            Method: HttpVerb.Post,
            Url: new Uri(_baseUrl + "/v1/sessions"),
            Headers: headers,
            Body: sig.SerializedBody);
        var resp = await _transport.SendAsync(req, ct).ConfigureAwait(false);
        if (resp.Status < 200 || resp.Status >= 300)
        {
            throw new SessionExchangeException(
                resp.Status,
                string.Format(
                    CultureInfo.InvariantCulture,
                    "SessionStore: POST /v1/sessions returned {0}: {1}",
                    resp.Status,
                    Truncate(resp.Body, 200)));
        }
        var envelope = JsonSerializer.Deserialize<BootstrapResponseEnvelope>(
            resp.Body, JsonOptions);
        var payload = envelope?.Data;
        if (payload is null ||
            string.IsNullOrEmpty(payload.SessionId) ||
            string.IsNullOrEmpty(payload.SessionSecret) ||
            string.IsNullOrEmpty(payload.ExpiresAt))
        {
            throw new SessionExchangeException(
                resp.Status,
                "SessionStore: response missing sessionId/sessionSecret/expiresAt");
        }
        if (!DateTimeOffset.TryParse(
            payload.ExpiresAt,
            CultureInfo.InvariantCulture,
            DateTimeStyles.AssumeUniversal | DateTimeStyles.AdjustToUniversal,
            out var expiresAt))
        {
            throw new SessionExchangeException(
                resp.Status,
                "SessionStore: response has invalid expiresAt");
        }
        return new Session(payload.SessionId, payload.SessionSecret, expiresAt);
    }

    private static string Truncate(string s, int max) =>
        s.Length <= max ? s : s.Substring(0, max);

    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNamingPolicy = null,
        PropertyNameCaseInsensitive = false,
    };

    private sealed record BootstrapResponseEnvelope(
        [property: JsonPropertyName("data")] BootstrapResponseBody? Data);

    private sealed record BootstrapResponseBody(
        [property: JsonPropertyName("sessionId")] string SessionId,
        [property: JsonPropertyName("sessionSecret")] string SessionSecret,
        [property: JsonPropertyName("expiresAt")] string ExpiresAt);

    public void Dispose() => _gate.Dispose();
}

/// <summary>
/// Transparent <see cref="ITransport"/> wrapper that signs every
/// outbound product request with the cached session and retries once
/// on 401 session_expired.
///
/// Wiring at the transport seam means every existing product method
/// (zyins, account, rapidsign, proxy) inherits auto-refresh without
/// per-method changes — they already go through ITransport.
/// </summary>
public sealed class SessionInterceptor : ITransport
{
    private readonly SessionStore _store;
    private readonly ITransport _inner;

    public SessionInterceptor(SessionStore store, ITransport inner)
    {
        if (store is null)
        {
            throw new ArgumentNullException(nameof(store));
        }
        if (inner is null)
        {
            throw new ArgumentNullException(nameof(inner));
        }
        _store = store;
        _inner = inner;
    }

    public async Task<TransportResponse> SendAsync(
        TransportRequest request, CancellationToken ct = default)
    {
        var resp = await SignAndSendAsync(request, ct).ConfigureAwait(false);
        if (!IsSessionExpired(resp))
        {
            return resp;
        }
        _store.Invalidate();
        return await SignAndSendAsync(request, ct).ConfigureAwait(false);
    }

    private async Task<TransportResponse> SignAndSendAsync(
        TransportRequest request, CancellationToken ct)
    {
        var sess = _store.CurrentSecret()
            ?? await _store.BootstrapAsync(ct).ConfigureAwait(false);
        var path = request.Url.PathAndQuery;
        var body = request.Body ?? string.Empty;
        var signed = SignRequest.Sign(
            method: request.Method.ToString().ToUpperInvariant(),
            path: path,
            body: body,
            sessionId: sess.SessionId,
            sessionSecret: sess.SessionSecret);
        var merged = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
        foreach (var kv in request.Headers)
        {
            merged[kv.Key] = kv.Value;
        }
        foreach (var kv in signed.AsDictionary())
        {
            merged[kv.Key] = kv.Value;
        }
        var signedReq = request with { Headers = merged };
        return await _inner.SendAsync(signedReq, ct).ConfigureAwait(false);
    }

    private static bool IsSessionExpired(TransportResponse resp)
    {
        if (resp.Status != 401)
        {
            return false;
        }
        if (!TryGetContentType(resp, out var ct) ||
            ct.IndexOf("json", StringComparison.OrdinalIgnoreCase) < 0)
        {
            return false;
        }
        try
        {
            using var doc = JsonDocument.Parse(resp.Body);
            return doc.RootElement.TryGetProperty("code", out var codeEl)
                && codeEl.ValueKind == JsonValueKind.String
                && codeEl.GetString() == "session_expired";
        }
        catch (JsonException)
        {
            return false;
        }
    }

    private static bool TryGetContentType(TransportResponse resp, out string value)
    {
        if (resp.Headers.TryGetValue("Content-Type", out var v) ||
            resp.Headers.TryGetValue("content-type", out v))
        {
            value = v;
            return true;
        }
        value = string.Empty;
        return false;
    }
}
